File: klinara-ios/klinara-ios/Services/Networking/APIRequest.swift
import Foundation

nonisolated enum HTTPMethod: String, Sendable {
    case get = "GET"
    case post = "POST"
    case patch = "PATCH"
    case put = "PUT"
    case delete = "DELETE"
}

/// Tek bir API çağrısının tarifi.
///
/// Servisler `URLRequest` kurmaz; yolu, gövdeyi ve kapsamı burada tarif eder,
/// başlıkları ve oturum yönetimini ``APIClient`` üstlenir. Bir başlığın tek bir
/// uçta unutulması bu sayede mümkün değil.
nonisolated struct APIRequest: Sendable {

    let method: HTTPMethod
    /// `API_PREFIX` (`/api/v1`) dahil DEĞİL — köke göreli: `services`, `staff/\(id)/services`.
    let path: String
    var query: [URLQueryItem] = []
    var body: (any Encodable & Sendable)?

    /// `false` ise `Authorization` başlığı gönderilmez ve 401'de yenileme denenmez.
    /// Yalnız `/auth/login`, `/auth/refresh` gibi public uçlar için.
    var requiresAuth = true

    /// Oturum token'ı yerine gönderilecek taşıyıcı.
    ///
    /// Giriş akışının ara token'ları (`mfa`, `tenant_select`) için gerekir:
    /// `/auth/2fa/setup` ve `/enable` challenge token'ını **gövdede değil**
    /// `Authorization` başlığında bekler. Verildiğinde 401'de yenileme denenmez —
    /// ara token'ın refresh'i yoktur, süresi dolduysa akış baştan başlar.
    var bearerOverride: String?

    /// Yan etkili POST'larda tekrar güvenliği. `POST /appointments` bunu
    /// destekler: aynı anahtarla ikinci istek yeni randevu üretmez, ilkinin
    /// yanıtını tekrar döndürür.
    var idempotencyKey: String?

    /// İyimser kilit. `PATCH /appointments/:id` ve `POST /:id/reschedule`
    /// bunu **zorunlu** kılar: başlık yoksa sunucu `428`, bayatsa `409
    /// VERSION_CONFLICT` döner. Değer ``weakETag(_:)`` ile kurulur — yanıt
    /// başlığını okumaya gerek yok, sürüm gövdedeki `version` alanında geliyor.
    var ifMatch: String?

    static func get(_ path: String, query: [URLQueryItem] = []) -> APIRequest {
        APIRequest(method: .get, path: path, query: query)
    }

    static func post(
        _ path: String,
        body: (any Encodable & Sendable)? = nil,
        idempotencyKey: String? = nil,
        ifMatch: String? = nil
    ) -> APIRequest {
        APIRequest(
            method: .post,
            path: path,
            body: body,
            idempotencyKey: idempotencyKey,
            ifMatch: ifMatch
        )
    }

    static func patch(
        _ path: String,
        body: (any Encodable & Sendable)? = nil,
        ifMatch: String? = nil
    ) -> APIRequest {
        APIRequest(method: .patch, path: path, body: body, ifMatch: ifMatch)
    }

    static func put(_ path: String, body: (any Encodable & Sendable)? = nil) -> APIRequest {
        APIRequest(method: .put, path: path, body: body)
    }

    /// `ifMatch`: `DELETE /package-definitions/:id` gibi iyimser kilit isteyen
    /// uçlar için — silme de bayat bir sürüm üzerinde yapılabilir.
    static func delete(_ path: String, ifMatch: String? = nil) -> APIRequest {
        APIRequest(method: .delete, path: path, ifMatch: ifMatch)
    }
}

/// Gövdesiz uçların (`204 No Content`) çözümleme hedefi.
nonisolated struct EmptyResponse: Decodable, Sendable {}

/// `{ "data": [...] }` sarmalayıcısı — liste uçlarının tamamı bu biçimde döner.
nonisolated struct ListEnvelope<Item: Decodable & Sendable>: Decodable, Sendable {
    let data: [Item]
}

/// Cursor sayfalamasının sayfa bilgisi — `common/pagination.ts` karşılığı.
nonisolated struct PageInfo: Decodable, Sendable, Equatable {
    let nextCursor: String?
    let hasMore: Bool
}

/// `{ "data": [...], "pageInfo": { … } }` — sayfalanan uçların zarfı.
///
/// ``ListEnvelope``'dan ayrı bir tip: cursor'u atan bir çağrı yeri sessizce
/// ilk sayfayla yetinirdi, bu ayrım derleyiciye söyletiyor.
nonisolated struct Page<Item: Decodable & Sendable>: Decodable, Sendable {
    let data: [Item]
    let pageInfo: PageInfo

    static var empty: Page<Item> {
        Page(data: [], pageInfo: PageInfo(nextCursor: nil, hasMore: false))
    }
}

/// `version` → `If-Match` başlığı. Sunucu `W/"3"` ve `"3"` biçimlerinin
/// ikisini de kabul ediyor; zayıf biçim `common/http/etag.ts`'in ürettiğiyle aynı.
nonisolated func weakETag(_ version: Int) -> String { "W/\"\(version)\"" }

File: klinara-android/app/src/main/java/com/klinara/android/services/networking/ApiRequest.kt
package com.klinara.android.services.networking

/** HTTP fiilleri. Sunucu sözleşmesinde kullanılanlar. */
enum class HttpMethod(val wire: String, val requiresBody: Boolean) {
    GET("GET", requiresBody = false),
    POST("POST", requiresBody = true),
    PATCH("PATCH", requiresBody = true),
    PUT("PUT", requiresBody = true),
    DELETE("DELETE", requiresBody = false),
}

/**
 * Bir istek tanımı. iOS `APIRequest` ile birebir aynı alanlar ve varsayılanlar —
 * iki istemcinin uç tanımları yan yana karşılaştırılabilir kalsın diye.
 *
 * [path] önek TAŞIMAZ: taban URL zaten `/api/v1` içeriyor (iOS `Info.plist`
 * `KlinaraAPIBaseURL` paritesi).
 */
data class ApiRequest(
    val method: HttpMethod,
    val path: String,
    val query: List<Pair<String, String>> = emptyList(),
    val body: RequestBodyPayload? = null,
    /** false ise `Authorization` ve `X-Branch-Id` hiç eklenmez. */
    val requiresAuth: Boolean = true,
    /**
     * `mfa` ara token'ı. Verilirse access token yerine bu kullanılır ve
     * `X-Branch-Id` **eklenmez** — ara token'ın şube kapsamı yoktur.
     */
    val bearerOverride: String? = null,
    val idempotencyKey: String? = null,
    /** Zayıf ETag değeri; [weakETag] ile üretilir. */
    val ifMatch: String? = null,
) {
    companion object {
        fun get(
            path: String,
            query: List<Pair<String, String>> = emptyList(),
            requiresAuth: Boolean = true,
        ) = ApiRequest(HttpMethod.GET, path, query = query, requiresAuth = requiresAuth)

        fun post(
            path: String,
            body: RequestBodyPayload? = null,
            requiresAuth: Boolean = true,
            bearerOverride: String? = null,
            idempotencyKey: String? = null,
            ifMatch: String? = null,
        ) = ApiRequest(
            HttpMethod.POST,
            path,
            body = body,
            requiresAuth = requiresAuth,
            bearerOverride = bearerOverride,
            idempotencyKey = idempotencyKey,
            ifMatch = ifMatch,
        )

        fun patch(
            path: String,
            body: RequestBodyPayload? = null,
            ifMatch: String? = null,
        ) = ApiRequest(HttpMethod.PATCH, path, body = body, ifMatch = ifMatch)

        fun put(
            path: String,
            body: RequestBodyPayload? = null,
        ) = ApiRequest(HttpMethod.PUT, path, body = body)

        fun delete(
            path: String,
            ifMatch: String? = null,
        ) = ApiRequest(HttpMethod.DELETE, path, ifMatch = ifMatch)

        /** `3` → `W/"3"` — sunucunun beklediği zayıf ETag biçimi. */
        fun weakETag(version: Int): String = "W/\"$version\""
    }
}

/**
 * Serileştirilmiş istek gövdesi.
 *
 * Gövde çağrı yerinde JSON'a çevrilir (`KlinaraJson.encode`), böylece [ApiClient]
 * generic bir `Encodable` kısıtı taşımak zorunda kalmaz ve reified tip parametresi
 * yalnız YANIT tarafında gerekir.
 */
@JvmInline
value class RequestBodyPayload(val json: String)
